File: fs/before_energy/719__find-k-th-smallest-pair-distance/cpp/find-k-th-smallest-pair-distance_benchmark_v2.cpp
#include <iostream>
template <class T>
static inline void DoNotOptimize(const T& value) {
    asm volatile("" : : "r,m"(value) : "memory");
}

#include <vector>
#include <algorithm>

class Solution {
public:
    // Time:  O(nlogn + nlogw), n = len(nums), w = max(nums)-min(nums)
    // Space: O(1)
    int smallestDistancePair(std::vector<int>& nums, int k) {
        // Sliding window solution with binary search over answer
        std::sort(nums.begin(), nums.end());
        int left = 0;
        int right = nums.back() - nums.front() + 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (possible(mid, nums, k)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

private:
    static bool possible(int guess, const std::vector<int>& nums, int k) {
        long long count = 0;
        int left = 0;
        for (int right = 0; right < static_cast<int>(nums.size()); ++right) {
            while (nums[right] - nums[left] > guess) {
                ++left;
            }
            count += right - left;
        }
        return count >= static_cast<long long>(k);
    }
};

int main() {
    struct Test {
        std::vector<int> nums;
        int k;
    };

    std::vector<Test> tests = {
        {{1, 100}, 1},
        {{1, 3, 1}, 1},
        {{1, 2, 3, 4, 5}, 4},
        {{10, 10, 10, 10}, 6},
        {{1, 6, 1, 2, 5, 9}, 7},
        {{0, 0, 0, 0, 0, 0}, 10},
        {{0, 1000000}, 1},
        {{1, 1, 2, 2, 3, 3, 4, 4}, 16},
        {{1, 3, 6, 10, 15, 21, 28, 36}, 20},
        {{4, 1, 4, 2, 8, 5, 7, 2, 3}, 12}
    };

    Solution solution;

    volatile int sink = 0;
    const int iterations = 1000;
    for (int iter = 0; iter < iterations; ++iter) {
        //int checksum = 0;
        for (const auto& t : tests) {
            std::vector<int> nums = t.nums;
            int result = solution.smallestDistancePair(nums, t.k);
            DoNotOptimize(result);
            //checksum += result;
        }
        //sink = checksum;
    }

    return 0;
}

File: fs/before_checksum/719__find-k-th-smallest-pair-distance/cpp/find-k-th-smallest-pair-distance_benchmark_v2.py
# Time:  O(nlogn + nlogw), n = len(nums), w = max(nums)-min(nums)
# Space: O(1)

class Solution(object):
    def smallestDistancePair(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        # Sliding window solution
        def possible(guess, nums, k):
            count, left = 0, 0
            for right, num in enumerate(nums):
                while num-nums[left] > guess:
                    left += 1
                count += right-left
            return count >= k

        nums.sort()
        left, right = 0, nums[-1]-nums[0]+1
        while left < right:
            mid = left + (right-left)/2
            if possible(mid, nums, k):
                right = mid
            else:
                left = mid+1
        return left

if __name__ == "__main__":
    tests = [
        {'nums': [1, 100], 'k': 1},
        {'nums': [1, 3, 1], 'k': 1},
        {'nums': [1, 2, 3, 4, 5], 'k': 4},
        {'nums': [10, 10, 10, 10], 'k': 6},
        {'nums': [1, 6, 1, 2, 5, 9], 'k': 7},
        {'nums': [0, 0, 0, 0, 0, 0], 'k': 10},
        {'nums': [0, 1000000], 'k': 1},
        {'nums': [1, 1, 2, 2, 3, 3, 4, 4], 'k': 16},
        {'nums': [1, 3, 6, 10, 15, 21, 28, 36], 'k': 20},
        {'nums': [4, 1, 4, 2, 8, 5, 7, 2, 3], 'k': 12}
    ]

    solution = Solution()

    sink = 0
    iterations = 1
    for _ in xrange(iterations):
        checksum = 0
        for t in tests:
            nums = list(t['nums'])
            result = solution.smallestDistancePair(nums, t['k'])
            checksum += result
        sink += checksum
    print(sink)